node/router/src/heartbeat.rs

```diff
@@ -104,8 +105,44 @@ pub trait Heartbeat<N: Network>: Outbound<N> {
         }
     }
 
-    /// This function removes the oldest connected peer, to keep the connections fresh.
-    /// This function only triggers if the router is above the minimum number of connected peers.
+    /// Returns a sorted vector of network addresess of all removable connected peers
+    /// where the first entry has the lowest priority andthe last one the highest.
+    ///
+    /// Rules:
+    ///     - Trusted peers and bootstrap nodes are not removable
+    ///     - Peers that we are currently syncing with are not removable
+    ///     - Validators are considered higher priority than provers or clients
+    ///     - Connections that have not been seen in a while are considered lower priority
+    fn get_removable_peers(&self) -> Vec<Peer<N>> {
+        // The trusted peers (specified at runtime)
+        let trusted = self.router().trusted_peers();
+        // The hardcoded bootstrap nodes
+        let bootstrap = self.router().bootstrap_peers();
+        // Are we synced already? (cache this here, so it does not need to be recomputed)
+        let is_block_synced = self.is_block_synced();
+
+        // Sort by priority, where lowest priority will be at the beginning
+        // of the vector.
+        // Note, that this gives equal priority to clients and provers, which
+        // we might want to change in the future
+        let mut peers = self.router().get_connected_peers();
+        peers.sort_by_key(|peer| (peer.is_validator(), peer.last_seen()));
+
+        // Deterimine which of the peers can be removed
+        peers
+            .into_iter()
+            .filter(|peer| {
+                !trusted.contains(&peer.ip()) // Always keep trusted nodes
+                  && !bootstrap.contains(&peer.ip()) // Always keep bootstrap nodes
+                  && !self.router().cache.contains_inbound_block_request(&peer.ip()) // This peer is currently syncing from us
```

Comment:
Can this be exploited in a way where a malicious validator just need to keep sending block requests to prevent the peer from disconnecting from them?

Reply:
Yes, same for malicious clients, this was already an issue before this PR.

@kaimast want to create a new issue for this? A simple solution [as noted here](https://github.com/ProvableHQ/snarkOS/issues/3026) would be to refresh a random selection of syncing peers, but I bet with further analysis we can avoid hurting sync performance of honest nodes.

Reply:
Even without this check, couldn't a malicious peer send Ping messages frequently to appear more responsive and receive higher priority?

Reply:
> Even without this check, couldn't a malicious peer send Ping messages frequently to appear more responsive and receive higher priority?

Oe nice one. We can think of some potential mitigations, e.g. reverting to `choose_multiple`, perhaps weighing clients heavier than validators, and banning when seeing very frequent ping messages (which we might already do).

You can add ideas in the new issue but I still think out of scope for this PR.

Reply:
Agreed, not in scope for this PR. Can be addressed in the future for changes directly related to handling malicious peers.